File: src/main/java/org/example/websitesalephone/dto/user/CreateUserDto.java
package org.example.websitesalephone.dto.user;

import lombok.Builder;
import lombok.Data;
import org.example.websitesalephone.entity.User;
import org.example.websitesalephone.enums.Gender;
import org.example.websitesalephone.utils.Constants;
import org.example.websitesalephone.utils.Utils;
import org.springframework.security.crypto.bcrypt.BCrypt;

import java.time.OffsetDateTime;
import java.util.UUID;

@Data
@Builder
public class CreateUserDto {

    private String id;

    private String loginId;

    private String fullName;

    private String email;

    private String telNo;

    private String note;

    private String role;

    private String address;

    private String gender;

    public static User toEntity(CreateUserDto dto) {
        User user = new User();

        user.setId(UUID.randomUUID().toString());
        user.setUsername(dto.getLoginId());
        user.setFullName(dto.getFullName() == null ? null : dto.getFullName().trim());
        user.setEmail(dto.getEmail());
        user.setDescription(dto.getNote());
        user.setPasswordHash(BCrypt.hashpw("666666", BCrypt.gensalt()));
        user.setPasswordExpiredAt(OffsetDateTime.now().plusDays(Constants.PASSWORD_EXPIRE_DAYS));
        user.setAddress(dto.getAddress());
        user.setGender(dto.getGender());
        user.setPhone(dto.getTelNo());

        return user;
    }
}

File: cms-user/src/models/CreateUserDto.ts
export interface IRoleDto {
    id?: string;
    name: string;
}

export interface ICreateUserDto {
    id?: string;
    loginId: string;
    fullName?: string;
    profileImg?: string;
    email?: string;
    telNo?: string;
    gender?: string;
    userCode?: string;
    note?: string;
    password: string;
    role?: string;
    address?: string;
    isDeleted?: boolean
}

export class CreateUserDto {
    public id?: string;
    public loginId: string;
    public fullName?: string;
    public profileImg?: string;
    public email?: string;
    public telNo?: string;
    public gender?: string;
    public userCode?: string;
    public note?: string;
    public password: string;
    public role?: string;
    public address?: string;
    public isDeleted?: boolean;

    constructor(data: {
        loginId: string;
        password: string;
        id?: string;
        fullName?: string;
        profileImg?: string;
        email?: string;
        telNo?: string;
        gender?: string;
        userCode?: string;
        note?: string;
        role?: string;
        address?: string;
        isDeleted?: boolean;
    }) {
        this.id = data.id?.trim();
        this.loginId = data.loginId.trim();
        this.fullName = data.fullName?.trim();
        this.profileImg = data.profileImg?.trim();
        this.email = data.email?.trim();
        this.telNo = data.telNo?.trim();
        this.gender = data.gender?.trim();
        this.userCode = data.userCode?.trim();
        this.note = data.note?.trim();
        this.password = data.password;
        this.role = data.role;
        this.address = data.address;
        this.isDeleted = data.isDeleted;
    }

    // Chuyển sang payload gửi API
    toPayload(): ICreateUserDto {
        return {
            id: this.id,
            loginId: this.loginId,
            fullName: this.fullName,
            profileImg: this.profileImg,
            email: this.email,
            telNo: this.telNo,
            gender: this.gender,
            userCode: this.userCode,
            note: this.note,
            password: this.password,
            role: this.role,
            address: this.address,
            isDeleted: this.isDeleted
        };
    }
}
